In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Project folder path as a string
path = r'/Users/Nimrat/Documents/Instacart Basket Analysis'

In [5]:
# Importing orders_products_merged_updated.pkl File
df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data' , 'ords_prods_merge_updated.pkl'))

In [6]:
#To check shape of the dataframe
df_ords_prods_merge.shape

(32404859, 18)

#2 - Find the aggregated mean of the “order_number” column grouped by “department_id” for entire dataframe.

In [7]:
#Grouping by department_id and calculating mean using agg() function on order_number
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The results for the subset and the full dataframe are different. The subset dataframe has lower average order numbers and full dataframe shows higher averages. Some departments shows large difference such as department 21 jumps from 22.90 in the subset to 25.5 in the full dataset.

#4 - Create a loyalty flag for existing customers using the transform() and loc() functions.

In [8]:
#Aggregating data with transform() function
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/yp/kt4_7p6531xbn6mhwvqy8y740000gp/T/ipykernel_78027/894939671.py:2: FutureWarning: The provided callable <function max at 0x105748ea0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [9]:
#To check the output
df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly busy        Average orders   
2   Mid-range product     Busiest day    Busiest days        Average orders   
3   Mid-range product  Regularly busy    Slowest days           Most orders   
4   Mid-range product 

In [10]:
#Create a loyalty flag using loc function
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [11]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [12]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
#To print frequenct of new loyalty flag column
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [14]:
#To check the output
df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly busy        Average orders   
2   Mid-range product     Busiest day    Busiest days        Average orders   
3   Mid-range product  Regularly busy    Slowest days           Most orders   
4   Mid-range product 

#5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [15]:
#To determine basic statistics
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#6 - The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [16]:
#Aggregating data with transform() function
df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/yp/kt4_7p6531xbn6mhwvqy8y740000gp/T/ipykernel_78027/689289674.py:2: FutureWarning: The provided callable <function mean at 0x1057498a0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [17]:
#To check results
df_ords_prods_merge[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0       138           6.935811
1       138           6.935811
2       709           7.930208
3       764           4.972414
4       764           4.972414
5       777           6.935398
6       825           5.957576
7       910           6.680000
8      1052           7.162500
9      1052           7.162500
10     1379           5.813636
11     1480           5.081633
12     1480           5.081633
13     1480           5.081633
14     1494           6.803797

In [18]:
#To create spending flag
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [19]:
#To print frequency of spending flag
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [20]:
#To check output of user_id, avg_product_price and spending flag
df_ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0       138           6.935811   Low spender
1       138           6.935811   Low spender
2       709           7.930208   Low spender
3       764           4.972414   Low spender
4       764           4.972414   Low spender
5       777           6.935398   Low spender
6       825           5.957576   Low spender
7       910           6.680000   Low spender
8      1052           7.162500   Low spender
9      1052           7.162500   Low spender
10     1379           5.813636   Low spender
11     1480           5.081633   Low spender
12     1480           5.081633   Low spender
13     1480           5.081633   Low spender
14     1494           6.803797   Low spender

#7 - In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [21]:
#Aggregating data with transform() function
df_ords_prods_merge['order_regularity_median'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/yp/kt4_7p6531xbn6mhwvqy8y740000gp/T/ipykernel_78027/528648907.py:2: FutureWarning: The provided callable <function median at 0x106994ea0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['order_regularity_median'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [22]:
#To check the output
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'order_regularity_median']].head(15)

user_id  days_since_prior_order  order_regularity_median
0       138                     3.0                      8.0
1       138                    20.0                      8.0
2       709                     6.0                      8.0
3       764                     NaN                      9.0
4       764                     9.0                      9.0
5       777                    26.0                     11.0
6       825                    30.0                     20.0
7       910                    30.0                      6.0
8      1052                    19.0                     10.0
9      1052                    15.0                     10.0
10     1379                     NaN                     20.0
11     1480                     7.0                      7.0
12     1480                     7.0                      7.0
13     1480                    12.0                      7.0
14     1494                     2.0                      7.0

In [23]:
#To create order frequency flag
df_ords_prods_merge.loc[df_ords_prods_merge['order_regularity_median'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['order_regularity_median'] <= 20) & (df_ords_prods_merge['order_regularity_median'] > 10), 'order_frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['order_regularity_median'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [24]:
#To print frequency of order_frequency_flag
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [25]:
#To check columns for the updated dataframe
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'order_regularity_median', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  order_regularity_median  \
0       138                     3.0                      8.0   
1       138                    20.0                      8.0   
2       709                     6.0                      8.0   
3       764                     NaN                      9.0   
4       764                     9.0                      9.0   
5       777                    26.0                     11.0   
6       825                    30.0                     20.0   
7       910                    30.0                      6.0   
8      1052                    19.0                     10.0   
9      1052                    15.0                     10.0   
10     1379                     NaN                     20.0   
11     1480                     7.0                      7.0   
12     1480                     7.0                      7.0   
13     1480                    12.0                      7.0   
14     1494                     2.0                      7.0   

   order_frequency_flag  
0     Frequent customer  
1     Frequent customer  
2     Frequent customer  
3     Frequent customer  
4     Frequent customer  
5      Regular customer  
6      Regular customer  
7     Frequent customer  
8     Frequent customer  
9     Frequent customer  
10     Regular customer  
11    Frequent customer  
12    Frequent customer  
13    Frequent customer  
14    Frequent customer

#9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [26]:
#Export dataframe in pickle format
df_ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merged_grouped.pkl'))